app/controllers/alchemy/admin/pages_controller.rb

```diff
@@ -25,6 +27,59 @@ def index
         end
       end
 
+      # Returns all pages as a tree from the root given by the id parameter
+      #
+      def tree
+        authorize! :tree, :alchemy_admin_pages
+
+        tree = []
+        path = [{id: @page.parent_id, children: tree}]
+        page_list = @page.self_and_descendants
+
+        page_list.each_with_index do |page, i|
+          if page.parent_id != path.last[:id]
+            if path.map { |o| o[:id] }.include?(page.parent_id) # Lower level
+              path.pop while path.last[:id] != page.parent_id
+            else # One level up
+              path << path.last[:children].last
+            end
+          end
+
+          has_children = page_list[i + 1] && page_list[i + 1].parent_id == page.id
+          level = path.count
+
+          path.last[:children] << {
```

Comment:
I've been working on [making this a serializer](https://gist.github.com/jorrizza/e9b4314a0c58c6be9d3a). It's not pretty. We have to instantiate a serializer for every page separately in that loop. This is slow and eats resources like crazy. The API call becomes about 30% slower this way. Also, the serializer has no access to CanCan or the loop itself. I had to (ab)use the metadata support to get that in. 

I agree this is not the right place for this bit of code. You mentioned making this a view? How did you have that in mind? My preferred solution would be to put this in either a helper (because it has access to the controller scope) or in a concern to include in `Alchemy::Page`.